LSPS5/README.md

```diff
@@ -0,0 +1,375 @@
+    [JSON-RPC 2.0 Notification Object][], which is the
+    webhook notification object.
+
+The notification delivery service, once it has received
+the timestamp, signature, and webhook notification object:
+
+* MUST check that the timestamp is within 10 minutes
+  (before or after) of its local time.
+* MUST remember the signature for at least 20 minutes,
+  and check that the current signature is not equal to
+  some remembered signature from a previous webhook
+  call.
+* MUST validate the signature against the message
+  template above, with the given timestamp and webhook
+  notification object.
+* MAY filter out particular notification methods by
+  its own policy.
+
+> **Rationale** The timestamp and signature checks are
+> intended to avoid replay attacks, where some third
+> party is able to copy the JSON object sent by the LSP
+> and replays it in order to frame the LSP for spamming
+> the client.
+
+The response to the webhook is unspecified.
+
+* For HTTPS webhooks (i.e. `POST`):
+  * The LSP MUST consider a `200 OK` response a
+    success, and MUST ignore the rest of the response,
+    including HTTP headers and the response body.
+  * If the response is not `200 OK`, the LSP SHOULD
+    log this as an unusual event, including the
+    HTTP headers and response body if any, and
+    MUST otherwise ignore the response.
+  * The LSP SHOULD NOT follow any redirection
+    responses.
+
+> **Rationale** As a notification, the user of the
+> client may ignore it, or disable notifications from
+> the client application entirely; thus, delivery of
+> the notification is never assured and it is pointless
+> to feed back anything to the LSP about whether the
+> notification was delivered to the notification
+> delivery service, the mobile OS server, or the client
+> mobile device.
+
+If the notification delivery service accepts the push
+notification, it MUST construct a suitable message to
+display to the user, and send that message, together
+with any required authentication and authorization, to
+the mobile OS server, for display as a push notification
+of the client.
+
+The client SHOULD process what is needed to
+react to the notification (e.g. connect to the LSP and
+accept the payment for a `lsps5.you_have_new_payments`
+notification) if the mobile environment allows it.
+
+#### Webhook Notification Methods
+
+The LSP MUST use one of the methods listed in this
+section for the webhook notification object.
+
+The LSP MUST NOT use any of these methods on the LSPS0
+interface; these notification methods are only for the
+webhook interface.
+
+Parameters are required unless otherwise stated.
+
+* `lsps5.you_have_just_registered` - The client has just
+  recently successfully called the `lsps5.set_webhook`
+  API.
+  No parameters `{}`.
+* `lsps5.you_have_new_payments` - The client has one or
+  more payments pending to be received.
+  Parameters:
+  * `total_msat` - The total amount, in millisatoshi,
```

Comment:
We are already talking about removing the `total_msat` due to concerns that the OS push notification can be surveilled --- sure we can recommend that the notification delivery service encrypt the data, but it is better to not have the data in the first place so that implementors do not mistakenly send it in the clear to the OS push notification service and cause surveillance issues.

There is also the idea that LSPs can limit spam by just sending this wakeup signal once even if multiple payments arrive; i.e. if the LSP POSTs to the webhook a "you have new payments" notification, and some more incoming payments arrive but the client has not come online yet, the LSP just waits it out until the client does come online and starts handling the payments.  Or is this kind of filtering/batching better left to the notification delivery service?

In theory, if the mobile device had enough time, it could connect to the LSP (which it has to do anyway to claim anything) and get the amount and the hash from the `update_add_htlc` message.  Just how limited is the ability of the client to get CPU time when backgrounded or the phone is locked? If the client cannot make an outgoing TCP/IP connection anyway, it cannot receive `update_add_htlc` or send `commitment_signed` or `ack_and_revoke` anyway, so the extra information might not be really useful?

Reply:
I agree total_amount is not needed. The mobile client has a very limited time to:
1. Start the node (and signer if exists).
2. Connect to the LSP peer.
3. Wait for the payment to be settled.

Without payment hash the mobile client would need to:
1. Wait for Some payment to be fulfilled or intercept the update_add_htlc message or poll for pending payments and then attempt (on best effort) to match the payment to the  notification so the app can display a proper message (e.g: "X sats have been received" )

Depends on the OS and the state of the device (battery, CPU, etc...) the OS gives the device some limited time, from a few seconds up to 30 seconds  to accomplish the background task so anything that can offload work from the device increases payment reliability. 
Is there anything wrong with sending payment hash?

Reply:
> Is there anything wrong with sending payment hash?

That kinda assumes that each payment is one notification. Indeed, as the name suggests, `total_msat` could imply that a single LSPS5 notification is multiple payments. Is a one-notification-per-payment the model desired? Why not just wake up and handle *all* the incoming HTLCs, then display an aggregate notification "XX sats have been received on NN payments"? Why is matching even necessary, why not just handle all the incoming HTLCs? In particular, aggregating multiple `update_add_htlc`s can be done and a single `commitment_signed`-`revoke_and_ack` turnaround is then done, saving total bytes sent on the TCP/IP tunnel.

An issue also is that in a PTLC world, the expectation is that, even at the hop from the penultimate to the ultimate payee, the point is blinded. The blinding factor is in the onion that goes to the ultimate payee.  In the LSP case, the penultimate node is the LSP and the ultimate payee is the client, meaning that the point that the LSP will send in a PTLC to the client is tweaked from the actual payment of the client, meaning the client has to go get the onion. Should we also include the onion in the notification? This would also help in `keysend`-like protocols, which would be needed for asynchronous receive anyway (the preimage in asyncpay is in the onion that goes to the client).

Reply:
Fair point. I am convienced we can remove the payment hash.

Reply:
ok